python_modules/dagster/dagster_tests/core_tests/structured_config_tests/test_structured_config_types.py

```diff
@@ -300,3 +302,136 @@ def a_job():
+        pet_type: Literal["dog"]
+        barks: float
+
+    class Lizard(Config):
+        pet_type: Literal["reptile", "lizard"]
+        scales: bool
+
+    class OpConfigWithUnion(Config):
+        pet: Union[Cat, Dog, Lizard] = Field(..., discriminator="pet_type")
+        n: int
+
+    executed = {}
+
+    @op
+    def a_struct_config_op(config: OpConfigWithUnion):
+        if config.pet.pet_type == "cat":
+            assert config.pet.meows == 2
+        elif config.pet.pet_type == "dog":
+            assert config.pet.barks == 3.0
+        elif config.pet.pet_type == "lizard":
+            assert config.pet.scales
+        assert config.n == 4
+
+        executed["yes"] = True
+
+    @job
+    def a_job():
+        a_struct_config_op()
+
+    assert a_job
+
+    a_job.execute_in_process(
+        {"ops": {"a_struct_config_op": {"config": {"pet": {"cat": {"meows": 2}}, "n": 4}}}}
+    )
+    assert executed["yes"]
+
+    executed = {}
+    a_job.execute_in_process(
+        {"ops": {"a_struct_config_op": {"config": {"pet": {"dog": {"barks": 3.0}}, "n": 4}}}}
+    )
+    assert executed["yes"]
+
+    executed = {}
+    a_job.execute_in_process(
+        {"ops": {"a_struct_config_op": {"config": {"pet": {"lizard": {"scales": True}}, "n": 4}}}}
+    )
+    assert executed["yes"]
+
+    executed = {}
+    a_job.execute_in_process(
+        {"ops": {"a_struct_config_op": {"config": {"pet": {"reptile": {"scales": True}}, "n": 4}}}}
+    )
+    assert executed["yes"]
+
+    # Ensure passing value which doesn't exist errors
+    with pytest.raises(DagsterInvalidConfigError):
+        a_job.execute_in_process(
+            {"ops": {"a_struct_config_op": {"config": {"pet": {"octopus": {"meows": 2}}, "n": 4}}}}
+        )
+
+    # Disallow passing multiple discriminated union values
+    with pytest.raises(DagsterInvalidConfigError):
+        a_job.execute_in_process(
+            {
+                "ops": {
+                    "a_struct_config_op": {
+                        "config": {
+                            "pet": {"reptile": {"scales": True}, "dog": {"barks": 3.0}},
+                            "n": 4,
+                        }
+                    }
+                }
+            }
+        )
+
+
+def test_nested_discriminated_unions():
+    class Poodle(Config):
+        breed_type: Literal["poodle"]
+        fluffy: bool
+
+    class Dachshund(Config):
+        breed_type: Literal["dachshund"]
+        long: bool
+
+    class Cat(Config):
+        pet_type: Literal["cat"]
+        meows: int
+
+    class Dog(Config):
+        pet_type: Literal["dog"]
+        barks: float
+        breed: Union[Poodle, Dachshund] = Field(..., discriminator="breed_type")
+
+    class OpConfigWithUnion(Config):
+        pet: Union[Cat, Dog] = Field(..., discriminator="pet_type")
+        n: int
+
+    executed = {}
+
+    @op
+    def a_struct_config_op(config: OpConfigWithUnion):
+        assert config.pet.breed.fluffy  # type: ignore[union-attr]
```

Comment:
let's have the test case check the discriminator so we don't need the type ignore